Tx Processor separation, Issue fixes, Embitz "news"

db4ple opened the pull request:
This is a huge change, unfortunately I had to reapply some of  changes from #1693 manually. I was not able to do this in a nicer way. However, changes still seem to be working. Although I had to adapt them to my changed structure.  I am sorry for  that.

- Embitz marked as unmaintained now (I made the bootloader for the mcHF build, but nothing else), Readme tells it now.
- Moved the I2C access out of the interrupt handler (should not be done there)
- Moved the transmit data processing out of the audio driver and cleaned it up. It should work as before,
  as now also BPSK can use the CW as input  source, RTTY now will be able to generate a sidetone on the mcHF, too (since we are transmitting at zero if (like PSK, CW and FreeDV).
- Cleaned up a lot of complicated things in terms of tx signal processing, especially the USB vs. LSB vs. freq shift things are now much better to understand and clearly separated. 
- Ultimate goal is to make RX/TX sample rates independent from IQ sample rates so that we can switch to 96k or higher.

For more information, please see the commit messages.

db4ple committed:
EmBitz builds marked as unsupported for now
Tried to align builds with Eclipse config. But EmBitz does
not support libc_nano for F7. Someone else should take it from here

Add comment that EmBitz builds are currently broken and are in search of
a maintainer to Readme.md

db4ple committed:
Moved Codec Restart out of Interrupt Handler

m-chichikalov commented:
Hi, Danilo, as I wrote, I found that it crashed when I tried to enable CW input for BPSK. Did you see the same and found why? Or nothing noticed?

db4ple commented:
No, it did not crash. It may have to do with my other changes. Although I would not know why.

To be honest, I did not test PSK reception on the other end, but the waveforms looked and sounded ok. But I was able to go on and off TX just like for RTTY.

m-chichikalov commented on mchf-eclipse/drivers/audio/audio_agc.c:
```diff
@@ -0,0 +1,567 @@
+    {
+        if (++agc_wdsp.out_index >= agc_wdsp.ring_buffsize)
+        {
+            agc_wdsp.out_index -= agc_wdsp.ring_buffsize;
```
There is a member field ring_buffsize. Seems like it is constant, init and never change later. Maybe get rid of it and use AGC_WDSP_RB_SIZE instead?

db4ple replied:
I don't know enough about the WDSP AGC to tell you if there is a reason for changing the size dynamically at some point. You have to ask @DD4WH for more info on this code or checkout the original WDSP library.

m-chichikalov replied:
Ok, thanks.

m-chichikalov commented:
> No, it did not crash. It may have to do with my other changes. Although I would not know why.
> 
> To be honest, I did not test PSK reception on the other end, but the waveforms looked and sounded ok. But I was able to go on and off TX just like for RTTY.

Ok it's building right now, I mean your branch, I'll try and let you know, maybe it's because of power_factor still 0 on my side...

m-chichikalov commented:
Tried BPSK, it works, no crashes.
But I again do not hear anything not when I press Tune, or when TX in CW,RTTY and BPSK.

db4ple commented:
No sidetone on a mcHF or an OVI40?

m-chichikalov commented:
OVI40

db4ple commented:
I think, I know why. I removed correctHalfWord in the Tx Sidetone Fill function after I moved the TxIqFinalProcessing down. But this is of course wrong, since audioDst != dst. And if both IQ and audio codec are using the 32bit format, we need it for both.

db4ple committed:
Audio Driver breakup (fist stage)

- moved out agc from main audio_driver.c
- removed OBSOLETE_NR and OBSOLETE_AGC code (which has not been in use
  for a while and has been replaced)
- moved tx code out of audio_driver.c into tx_processor.c
- worked on making tx functions less directly accessing the global buffers
 (instead getting them as parameters)
- simplified sideband & freq shift use. Now each modulator generates correct iq
  which can be passed through the final iq processing without IQ swap.

User improvements:
- fixed an issue with the twinpeaks calling Codec_RestartI2S in the interrupt handler
- fixed BPSK transmit to correctly transmit at selected sideband
- BPSK can now be entered using a cw paddle (like RTTY).
- RTTY now generates a sidetone on mcHF
- fixed wrong state buffer allocation for FreeDV RX IIR filter

Please test all tx modes. RX should be unaffected, no significant changes made in this area.

db4ple commented:
Ok, now the sidetone is back. I find the mixture between CW "feedback" and the generated signal difficult to manage. We'll probably have to be a little smarter here. But anyway, you'll be able to hear it again.

m-chichikalov commented:
> Ok, now the sidetone is back. I find the mixture between CW "feedback" and the generated signal difficult to manage. We'll probably have to be a little smarter here. But anyway, you'll be able to hear it again.

Do you want me to check? if so, will do it tomorrow.
